packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py

```diff
@@ -61,11 +62,23 @@ async def _file_chunk_reader(
             yield chunk
 
 
-async def _file_chunk_writer(file: Path, response: ClientResponse, pbar: tqdm):
+@runtime_checkable
+class LogRedirectCB(Protocol):
+    async def __call__(self, msg: str) -> None:
+        ...
+
+
+async def _file_chunk_writer(
+    file: Path,
+    response: ClientResponse,
+    pbar: tqdm,
+    io_log_redirect_cb: Optional[LogRedirectCB],
+):
     async with aiofiles.open(file, "wb") as file_pointer:
         while chunk := await response.content.read(CHUNK_SIZE):
             await file_pointer.write(chunk)
-            pbar.update(len(chunk))
+            if pbar.update(len(chunk)) and io_log_redirect_cb:
```

Comment:
Q: so ``pbar`` always prints in stdout right? regardless of ``io_log_redirect_cb``?

Reply:
ok, so there are 2 things. pbar.update returns True if it should actually print (cause it does not always print based on some internal rationale like minimum amount of iterations, time, etc etc...).
Then you can also get all the progress info as a dictionary. or as how it prints it by itself.
Also I think that tqdm prints to stderr

Reply:
fyi on this topic: RabbitMQ is thought to be able to handle approx. <50k msg/sec

Reply:
concerning the througput. these are logs. treated as logs are. I guess nobody ever checked if rabbitmq is under/over used. something to chec at some point.
and I tend to disagree with saying that progress is not critical. this is user feedback and is very important.